src/dso_api/dynamic_api/permissions.py

```diff
@@ -1,99 +1,25 @@
-
-    # Make sure that names are snake_cased
-    # if used for a lookup in models.DatasetTable.objects
-    # TODO: do snake_case transformations at one centralized place to
-    # to be used in code (as much as possible)
-    dataset_id = to_snake_case(dataset_id)
-    table_id = to_snake_case(table_id)
-
-    try:
-        table = models.DatasetTable.objects.get(name=table_id, dataset__name=dataset_id)
-    except models.DatasetTable.DoesNotExist:
-        return TableScopes()
-
-    return TableScopes(
-        table=frozenset(level.auth for level in [table, table.dataset] if level.auth is not None),
-        fields={field.name: field.auth for field in table.fields.all() if field.auth is not None},
-    )
-
-
-@ttl_cache(ttl=60 * 60)
-def fetch_scopes_for_model(model: Type[Model]) -> TableScopes:
-    """ Get the scopes for a Django model, based on the Amsterdam schema information """
-
-    # If it is not a DSO-based model, we leave it alone
-    if not hasattr(model, "_dataset_schema"):
-        return TableScopes()
-    table = model._table_schema.id
-    dataset = model._meta.app_label
-
-    return fetch_scopes_for_dataset_table(dataset, table)
-
-
-def get_unauthorized_fields(request, model) -> Set[str]:
-    """Returns a list of field names to be excluded from a response."""
-    # is_authorized_for is added by authorization_django middleware
-    # only if an authorization check is necessary.
-    if not hasattr(request, "is_authorized_for"):
-        return set()
-
-    model_scopes = fetch_scopes_for_model(model)
-    unauthorized_fields = set()
-
-    for model_field in model._meta.get_fields():
-        required = model_scopes.table
-        field_scope = model_scopes.fields.get(model_field.name)
-        if field_scope is not None:
-            required = required | set([field_scope])
-
-        if not request.is_authorized_for(*required):
-            permission_key = get_permission_key_for_field(model_field)
-            if not request_has_permission(request=request, perm=permission_key):
-                unauthorized_fields.add(toCamelCase(model_field.name))
-
-    return unauthorized_fields
-
-
 def filter_unauthorized_expands(
-    request, expanded_fields: List[EmbeddedFieldMatch], skip_unauth=False
+    user_scopes: UserScopes, expanded_fields: List[EmbeddedFieldMatch], skip_unauth=False
 ) -> List[EmbeddedFieldMatch]:
     """Remove expanded fields if these are not accessible"""
-    if not hasattr(request, "is_authorized_for"):
-        # Not using 'authorization_django' middleware.
-        return expanded_fields
-
     result = []
     for match in expanded_fields:
-        scopes = fetch_scopes_for_model(match.field.related_model)
-        if not request.is_authorized_for(*scopes.table):
+        field_schema = match.field.field_schema
+        related_table = match.field.related_model.table_schema()
+
+        if not user_scopes.has_field_access(field_schema) or (
```

Comment:
Maybe use `not any(user_scopes.has_field_access(field_schema), user_scopes.has_table_access(related_table))` double negative logic is harder to grasp and often leads to logical mistakes in future.

Reply:
`any` needs an iterable though. We'd need to evaluate both checks. We can do

```python
not (user_scopes.has_field_access(field_schema) and user_scopes.has_table_access(related_table))
```

Reply:
Brainfart code ;)

```
not any([user_scopes.has_field_access(field_schema), user_scopes.has_table_access(related_table)])
```

Code in PR requires either of permissions to access field, one @lbam suggesting will require both.

Reply:
The check is `not A or not B`, which is `not (A and B)` or `not all([A,B])` or `any([not A, not B])`.

What I think might be more readable is to do both checks anyway, just to get the logical formula on a single line:

```python
field_access = user_scopes.has_field_access(match.field.field_schema)
table_access = user_scopes.has_table_access(match.field.related_model.table_schema())

if not (field_access and table_access):
    # ...
```

Reply:
Or something like this?

```
    for match in expanded_fields:
        field_perm = user_scopes.has_field_access(match.field.field_schema)
        table_perm = user_scopes.has_table_access(match.field.related_model.table_schema())
        if field_perm and table_perm:
            # Only add the field when there is permission
            result.append(match)
        else:
```

I dislike creating a temporary list with `all([...])` 

I'm also fine with `if not field_perm or not table_perm` as that was quite clear in my idea.

Reply:
The less negative logic - the better imo.

Reply:
Fine with me!

Reply:
aangepast!